File: Stark/API/Task.swift
import AppKit
import JavaScriptCore

@objc
protocol TaskJSExport: JSExport {
    init(path: String, arguments: [String]?, callback: JSValue)

    func terminate()
}

open class Task: Handler, TaskJSExport, HashableJSExport {
    var task: Process

    public required init(path: String, arguments: [String]?, callback: JSValue) {
        task = Process()
        task.launchPath = path

        if let arguments = arguments {
            task.arguments = arguments
        }

        super.init()

        task.terminationHandler = { [weak self] _ in
            self?.taskDidTerminate()
        }

        manageCallback(callback)
        launch()
    }

    func launch() {
        task.launch()
    }

    func terminate() {
        task.terminate()
    }

    @objc
    func taskDidTerminate() {
        perform(#selector(Task.callWithArguments(_:)), on: Thread.main, with: self, waitUntilDone: false, modes: nil)
    }
}

File: Stark/Resources/JavaScript/task.js
(function setup(Task) {
  const tasks = {};

  Task.run = (path, args, callback) => {
    const task = new Task(path, args, (handler) => {
      if (callback) {
        callback(handler);
      }

      Task.terminate(handler.id);
    });

    if (task) {
      tasks[task.id] = task;
      return task.id;
    }

    return null;
  };

  Task.terminate = (identifier) => {
    const task = tasks[identifier];

    if (task) {
      task.terminate();
      delete tasks[identifier];
    }
  };
}(Task));
